Kluwer Arbitration.js

```diff
@@ -0,0 +1,1011 @@
+			});
+		}
+	}*/
+
+	if (pubDetails && type == 'bookSection') {
+		// Attempting to get page range for bookSection
+		//let sectionNodes = pubDetails.tocNodes.flatMap((node) => (node.children ? node.children : node));
+		let sectionNodes = flattenNodes(pubDetails.tocNodes);
+		Z.debug(sectionNodes);
+		let section = sectionNodes.find(node => (node.docId.toLowerCase() == documentID.toLowerCase()));
+		if (section) Z.debug(section);
+		if (section.pageRange.first && section.pageRange.last) item.pages = section.pageRange.first + "-" + section.pageRange.last;
+	}
+	else if (type == 'journalArticle' | type == 'case') {
+		let volumeRegex = /Volume\s([a-zA-Z0-9]+)/;
+		let issueRegex = /Issue\s([a-zA-Z0-9]+)/;
+		let pagesRegex = /pp\.\s([a-zA-Z0-9]+)\s-\s([a-zA-Z0-9]+)/;
+
+		let volume = docDetails.BibliographicReference.match(volumeRegex);
+		let issue = docDetails.BibliographicReference.match(issueRegex);
+		let pages = docDetails.BibliographicReference.match(pagesRegex);
+
+		if (volume) item.volume = volume[1];
+		if (issue) item.issue = issue[1];
+		if (pages) {
+			if (type == 'case') item.firstPage = pages[1];
+			else item.pages = pages[1] + "-" + pages[2];
+		}
+	}
+
+	//item.url = url;
+
+	// Getting PDF
+	if (type == 'book') {
+		Z.debug("Before search: " + item);
+
+		var search = Zotero.loadTranslator("search");
+
+		search.setHandler("translators", function (obj, translators) {
+			search.setTranslator(translators);
+			search.translate();
+		});
+
+		search.setHandler("itemDone", function (obj, item) {
+			// Getting PDFs of all sections
+			let sectionNodes = flattenNodes(pubDetails.tocNodes);
+			let sectionList = sectionNodes.map(node => ([node.docId, node.label]));
+			let sections = Object.fromEntries(sectionList);
+			Z.debug(sections);
+			Z.selectItems(sections, function (items) {
+				if (!items) {
+					// CANCEL ITEM
+					return;
+				}
+
+				for (const [sectionID, sectionLabel] of Object.entries(items)) {
+					let pdfURL = "https://www.kluwerarbitration.com/document/print?title=PDF&ids=" + sectionID;
+					Z.debug(pdfURL);
+					item.attachments.push({
+						title: sectionLabel,
+						mimeType: "application/pdf",
+						url: pdfURL
+					});
+				}
+			});
+
+			item.attachments.push({
+				url: url,
+				title: "Read on Kluwer Arbitration",
+				mimeType: "text/html",
+				snapshot: false
+			});
+
+			item.complete();
+		});
+
+		search.setSearch(item);
+		// look for translators for given item
+		search.getTranslators();
+	}
+	else {
+		let pdfURL = '';
+		if (type == 'case' && docDetails.PublicationType == 'internet') pdfURL = "https://www.kluwerarbitration.com/document/GetPdf/" + docDetails.Id;
+		else pdfURL = "https://www.kluwerarbitration.com/document/print?title=PDF&ids=" + docDetails.Id;
+		Z.debug(pdfURL);
+		item.attachments.push({
+			title: "Full Text PDF",
+			mimeType: "application/pdf",
+			url: pdfURL
+		});
+
+		item.attachments.push({
+			url: url,
+			title: "Read on Kluwer Arbitration",
```

Comment:
Will rename to something like `Kluwer Arbitration Link`.

I initially saved the URL with `item.url` but I'm not sure it's semantically correct to give a URL to something that isn't inherently online (like a book or journal). That's why I chose to add the link as an attachment. But if established practice is to always set the item's URL field regardless of the type, I'll change it.

Reply:
Opted to set the URL on the items instead of a snapshot as the snapshots don't bring that much in the end (needs live javascript for the footnotes to be improved, which is a shame).